codegen/builtin_fs.go

```diff
@@ -742,10 +769,81 @@ func (dp DockerPush) Call(ctx context.Context, cln *client.Client, ret Register,
 	return ret.Set(fs)
 }
 
-var (
-	dockerCli  *command.DockerCli
-	dockerOnce sync.Once
-)
+func pushWithMoby(ctx context.Context, dockerAPI *DockerAPIClient, ref string, l progress.SubLogger) error {
+	creds, err := imagetools.RegistryAuthForRef(ref, dockerAPI.Auth)
+	if err != nil {
+		return err
+	}
+
+	rc, err := dockerAPI.ImagePush(ctx, ref, types.ImagePushOptions{
+		RegistryAuth: creds,
+	})
+	if err != nil {
+		return err
+	}
+
+	started := map[string]*client.VertexStatus{}
+
+	defer func() {
+		for _, st := range started {
+			if st.Completed == nil {
+				now := time.Now()
+				st.Completed = &now
+				l.SetStatus(st)
+			}
+		}
+	}()
+
+	dec := json.NewDecoder(rc)
+	var parsedError error
+	for {
+		var jm jsonmessage.JSONMessage
+		if err := dec.Decode(&jm); err != nil {
+			if parsedError != nil {
+				return parsedError
+			}
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if jm.ID != "" {
+			id := "pushing layer " + jm.ID
+			st, ok := started[id]
+			if !ok {
+				if jm.Progress != nil || jm.Status == "Pushed" {
+					now := time.Now()
+					st = &client.VertexStatus{
+						ID:      id,
+						Started: &now,
+					}
+					started[id] = st
+				} else {
+					continue
+				}
```

Comment:
Why not:

```suggestion
				if jm.Progress == nil && jm.Status != "Pushed" {
					continue
				}
```

instead of if/else...